File: day24_part1_2019.php

<?php

const Side = 5;
const Square = Side * Side;

function parse()
{
    $res = array_fill(0, Square, false);

    $file = fopen("input.txt", "r");
    if ($file === false) {
        die("Error opening file");
    }

    $row = 0;
    while (($line = fgets($file)) !== false) {
        for ($col = 0; $col < Side; $col++) {
            if ($line[$col] === '#') {
                $res[$row * Side + $col] = true;
            }
        }
        $row++;
    }

    fclose($file);

    return $res;
}

function next1($grid)
{
    $newGrid = array_fill(0, Square, false);

    for ($i = 0; $i < Square; $i++) {
        $row = (int)($i / Side);
        $col = $i % Side;
        $neighbours = 0;

        if ($row > 0 && $grid[$i - Side]) {
            $neighbours++;
        }
        if ($row < Side - 1 && $grid[$i + Side]) {
            $neighbours++;
        }
        if ($col > 0 && $grid[$i - 1]) {
            $neighbours++;
        }
        if ($col < Side - 1 && $grid[$i + 1]) {
            $neighbours++;
        }

        if ($grid[$i] && $neighbours != 1) {
            $newGrid[$i] = false;
            continue;
        }

        if (!$grid[$i] && ($neighbours == 1 || $neighbours == 2)) {
            $newGrid[$i] = true;
            continue;
        }

        $newGrid[$i] = $grid[$i];
    }

    return $newGrid;
}

function biodiversity($grid)
{
    $bio = 0;
    for ($i = 0; $i < Square; $i++) {
        if ($grid[$i]) {
            $bio += 1 << $i;
        }
    }
    return $bio;
}

$appeared = [];
$grid = parse();
$appeared[serialize($grid)] = true;

while (true) {
    $grid = next1($grid);
    if (isset($appeared[serialize($grid)])) {
        echo biodiversity($grid) . PHP_EOL;
        break;
    }
    $appeared[serialize($grid)] = true;
}

File: objectivec/day24_part1_2019.m
#import <Foundation/Foundation.h>

#define SIDE 5
#define SQUARE (SIDE * SIDE)

NSArray<NSNumber *> *parse() {
    NSMutableArray<NSNumber *> *res = [NSMutableArray arrayWithCapacity:SQUARE];
    NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
    NSArray *lines = [input componentsSeparatedByString:@"\n"];
    
    for (NSString *line in lines) {
        for (NSUInteger col = 0; col < SIDE; col++) {
            [res addObject:@([line characterAtIndex:col] == '#')];
        }
    }
    return res;
}

NSArray<NSNumber *> *next1(NSArray<NSNumber *> *grid) {
    NSMutableArray<NSNumber *> *newGrid = [NSMutableArray arrayWithCapacity:SQUARE];
    
    for (NSUInteger i = 0; i < SQUARE; i++) {
        NSUInteger row = i / SIDE, col = i % SIDE;
        NSUInteger neighbours = 0;

        if (row > 0 && [grid[i - SIDE] boolValue]) neighbours++;
        if (row < SIDE - 1 && [grid[i + SIDE] boolValue]) neighbours++;
        if (col > 0 && [grid[i - 1] boolValue]) neighbours++;
        if (col < SIDE - 1 && [grid[i + 1] boolValue]) neighbours++;

        if ([grid[i] boolValue] && neighbours != 1) {
            [newGrid addObject:@(NO)];
        } else if (![grid[i] boolValue] && (neighbours == 1 || neighbours == 2)) {
            [newGrid addObject:@(YES)];
        } else {
            [newGrid addObject:grid[i]];
        }
    }
    
    return newGrid;
}

NSInteger biodiversity(NSArray<NSNumber *> *grid) {
    NSInteger bio = 0;
    for (NSUInteger i = 0; i < SQUARE; i++) {
        if ([grid[i] boolValue]) bio += 1 << i;
    }
    return bio;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSMutableSet *appeared = [NSMutableSet set];
        NSArray<NSNumber *> *grid = parse();
        [appeared addObject:grid];

        while (YES) {
            grid = next1(grid);
            if ([appeared containsObject:grid]) {
                NSLog(@"%ld", (long)biodiversity(grid));
                break;
            }
            [appeared addObject:grid];
        }
    }
    return 0;
}